File: inst/tinytest/test_cats.R

set.seed(0)
r <- rast(nrows=10, ncols=10)
values(r) <- sample(3, ncell(r), replace=TRUE)
r0 <- r * 1

# categories in a single layer
lv <- data.frame(id=1:4, cover=c("forest", "water", "urban", "other"), stringsAsFactors=FALSE)
levels(r) <- lv
names(r) <- "land cover"
v <- cats(r)[[1]]
#coltab(r) <- rainbow(4)

expect_equal(v$id, lv[,1])
expect_equal(v$cover, lv[,2])

r <- droplevels(r)
expect_equal(levels(r)[[1]][,2], c("forest", "water", "urban"))
lv <- lv[-4,]

# reading/writing categories 
ftmp <- tempfile(fileext = ".tif")
z <- writeRaster(r, ftmp, overwrite=TRUE)
v <- cats(z)[[1]]
expect_equal(v$value, 1:3)
expect_equal(v$`land cover`, lv[,2])

# reading/writing subset of categories 
levels(r) = cats(r)[[1]][2:3,]
zz = writeRaster(r, ftmp, overwrite=TRUE)
v <- cats(zz)[[1]]

expect_equal(v$value, 2:3)
expect_equal(v$cover, lv[-1,2])

# categories in multiple layers
r2 <- rast(list(a=r0, b=r0))

# values are numeric initially
expect_equal(values(r2)[1,], c(a = 2, b = 2))

# set layer a categories
levels(r2) <- lv
expect_equal(levels(r2), list(lv, ""))

# verify samples for layer a have categories
expect_equal( r2[c(15,3)], data.frame(cover = factor(c("forest", "urban"), levels = lv[,2]), 
                                           b = c(1, 3), stringsAsFactors=FALSE ))

# set all layer categories
levels(r2) <- rep(list(lv), 2)
expect_equal(levels(r2), list(lv, lv))

# verify samples for layer a and layer b have categories
expect_equal(r2[c(15,3)], data.frame(cover = factor(c("forest", "urban"), levels = lv[,2]),
                                           cover = factor(c("forest", "urban"), levels = lv[,2]), stringsAsFactors=FALSE, check.names = FALSE))

# make sure no errors when show()ing factors
expect_silent(show(r2))

# catalyze multi-column lookup
r <- rast(nrows=10, ncols=10, vals=sample(1:5, 100, replace=TRUE))
levels(r) <- data.frame(ID=1:5, val1=101:105, val2=seq(0.1, 0.5, length.out=5))
r_cat <- catalyze(r)
expect_equal(nlyr(r_cat), 2)
expect_equal(as.vector(values(r_cat[[1]])), as.vector(values(r)) + 100)
expect_equal(as.vector(values(r_cat[[2]])), as.vector(values(r)) * 0.1)

File: python/tests/test_cats.py
"""
Tests ported from inst/tinytest/test_cats.R

Covers: levels / categories on SpatRaster.
"""
import numpy as np
import pandas as pd
import pytest
import geospat as pt
from geospat.rast import rast
from geospat.values import set_values
from geospat.levels import (
    levels, set_levels,
    cats, drop_levels, catalyze,
)


def _vals(r):
    return np.array(r.readValues(0, r.nrow(), 0, r.ncol()), dtype=float)


def make_r():
    np.random.seed(0)
    r = rast(nrows=10, ncols=10)
    v = np.random.choice([1, 2, 3], size=100).astype(float)
    return set_values(r, v)


def test_cats_roundtrip():
    """Categories set with set_levels() can be retrieved with cats()."""
    r = make_r()
    lv = pd.DataFrame({"id": [1, 2, 3, 4],
                       "cover": ["forest", "water", "urban", "other"]})
    r = set_levels(r, lv)
    v = cats(r)[0]
    assert list(v["id"]) == [1, 2, 3, 4]
    assert list(v["cover"]) == ["forest", "water", "urban", "other"]


def test_drop_levels():
    """drop_levels() removes categories not present in the data."""
    r = make_r()
    lv = pd.DataFrame({"id": [1, 2, 3, 4],
                       "cover": ["forest", "water", "urban", "other"]})
    r = set_levels(r, lv)
    r = drop_levels(r)
    remaining = cats(r)[0]
    # values are 1,2,3 only → 'other' (id=4) should be removed
    assert 4 not in list(remaining["id"])
    assert "other" not in list(remaining.iloc[:, 1])


def test_catalyze_nlyr():
    """catalyze() converts a factor raster to numeric layers."""
    np.random.seed(1)
    r = rast(nrows=10, ncols=10)
    r = set_values(r, np.random.choice(np.arange(1, 6), size=100).astype(float))
    lv = pd.DataFrame({"ID": range(1, 6),
                       "val1": range(101, 106),
                       "val2": [v * 0.1 for v in range(1, 6)]})
    r = set_levels(r, lv)
    r_cat = catalyze(r)
    assert r_cat.nlyr() == 2


def test_catalyze_layer_names():
    np.random.seed(1)
    r = rast(nrows=10, ncols=10)
    r = set_values(r, np.random.choice(np.arange(1, 6), size=100).astype(float))
    lv = pd.DataFrame({"ID": range(1, 6),
                       "val1": range(101, 106),
                       "val2": [v * 0.1 for v in range(1, 6)]})
    r = set_levels(r, lv)
    r_cat = catalyze(r)
    names = list(r_cat.names)
    assert "val1" in names
    assert "val2" in names


def test_catalyze_val1_values():
    """val1 layer should equal original integer values + 100."""
    np.random.seed(1)
    r = rast(nrows=10, ncols=10)
    v_orig = np.random.choice(np.arange(1, 6), size=100).astype(float)
    r = set_values(r, v_orig)
    lv = pd.DataFrame({"ID": range(1, 6),
                       "val1": range(101, 106),
                       "val2": [v * 0.1 for v in range(1, 6)]})
    r = set_levels(r, lv)
    r_cat = catalyze(r)
    # Retrieve val1 layer (layer 0)
    opt = pt.SpatOptions()
    r_val1 = r_cat.subset([0], opt)
    v_cat1 = _vals(r_val1)
    # Re-read the original values (may have been modified by set_levels)
    v_r = _vals(r)
    np.testing.assert_array_equal(v_cat1, v_r + 100)


